### PR TITLE
Add custom auth errors

### PR DESCRIPTION
**Pull request by yathindrakodithuwakku**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Review by brionmario — APPROVED** [review]

File: lib/src/authenticate.tsx [hunk 1]
```diff
@@ -156,21 +174,18 @@ const AuthProvider: FunctionComponent<PropsWithChildren<AuthProviderPropsInterfa
...
-                        });
+                    try{
+                        await signIn(
+                            { callOnlyOnRedirect: true }, 
+                            authParams?.authorizationCode, 
+                            authParams?.sessionState,
+                            authParams?.state
+                        );
+                    } catch(error) {
+                        if(error?.code === "JS-CRYPTO_UTILS-IVIT-IV02") {
```
Comment (brionmario) [inline lib/src/authenticate.tsx:185]: Better if these error codes are exported from `auth-js` SDK. We can track an issue for this.
```suggestion
                        if(error?.code === "JS-CRYPTO_UTILS-IVIT-IV02") {
```

File: lib/src/constants.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export const INVALID_SYSTEM_TIME = "INVALID_SYSTEM_TIME";
```
Comment (brionmario) [inline lib/src/constants.ts:19]:
```suggestion
export const INVALID_SYSTEM_TIME: string = "INVALID_SYSTEM_TIME";
```

File: lib/src/authenticate.tsx [hunk 3]
```diff
@@ -111,6 +125,10 @@ const AuthProvider: FunctionComponent<PropsWithChildren<AuthProviderPropsInterfa
         return AuthClient.on(hook, callback);
     };
     const trySignInSilently = () => AuthClient.trySignInSilently(state, dispatch);
+    
+    const [ errors, setErrors ] = useState([]);
```
Comment (brionmario) [inline lib/src/authenticate.tsx:129]: Is it possible to type this from the exception type thrown from the low level SDK?
```suggestion
    const [ errors, setErrors ] = useState<AuthError[]>([]);
```

**Comment by brionmario** [comment]
@thivi Could you review this as well?

**Review by brionmario — APPROVED** [review]

File: lib/src/authenticate.tsx [hunk 4]
```diff
@@ -111,6 +125,10 @@ const AuthProvider: FunctionComponent<PropsWithChildren<AuthProviderPropsInterfa
         return AuthClient.on(hook, callback);
     };
     const trySignInSilently = () => AuthClient.trySignInSilently(state, dispatch);
+    
+    const [ errors, setErrors ] = useState<AsgardeoAuthException[]>([]);
+
+    const addError = useCallback((error: AsgardeoAuthException) => setErrors([...errors, error]), []);
```
Comment (thivi) [inline lib/src/authenticate.tsx:131]: Why do we need a callback here? Can't we call `setErrors` directly?
I think we can do something like this on L186:
```
setErrors(errors => [ ...errors, error ]);
```
And any reason why we append errors to an array?

File: lib/src/authenticate.tsx [hunk 5]
```diff
@@ -23,14 +23,16 @@ import {
...
+    SignInConfig,
+    AsgardeoAuthException
 } from "@asgardeo/auth-spa";
 import { SPACustomGrantConfig } from "@asgardeo/auth-spa/src/models/request-custom-grant";
 import React, {
     FunctionComponent,
     PropsWithChildren,
     ReactNode,
     createContext,
+    useCallback,
```
Comment (thivi) [inline lib/src/authenticate.tsx:35]: We can remove this, right?

**Review by thivi — APPROVED** [review]

File: lib/src/authenticate.tsx (hunk 5 above)
Comment (yathindrakodithuwakku) [reply lib/src/authenticate.tsx]: done

**Pull request merged**